core/src/main/java/hudson/model/Computer.java

```diff
@@ -164,6 +164,13 @@
     private volatile String cachedHostName;
     private volatile boolean hostNameCached;
 
+    /**
+     * @see #getEnvironment()
+     */
+    private volatile EnvVars cachedEnvironment;
+    private volatile boolean environmentCached;
```

Comment:
Yes it is also an incorrect pattern. The question with cachedHostName is whether null implies that the host name was cached or whether it was tried and failed. 

If you have dual meaning from the `null` return value then you need to have a separate object state or just use a fricking lock.